### PR TITLE
unify the FV3-HRRR and FV3-HRRR-smoke

### PR DESCRIPTION
**Pull request by haiqinli**
1). The codes of ufs-weather-model/fv3atm/ccpp-physics is based on RRFS_dev for baseE.

2). The FV3_HRRR and FV3_HRRR_smoke are unified as one suite of FV3_HRRR. If the input namelist option rrfs_sd is false, it runs as the standard alone RRFS. IF rrfs_sd is true, it runs with smoke and dust for RRFS-SD. The meteorological fields are identical between the runs with rrfs_sd=T or rrfs_sd=F.

3). There is just one field_table, which contains smoke and dust tracers. We use the rrfs_sd option to control the tracers index. If rrfs_sd is false, only the default tracers of RRFS are read in. If rrfs_sd is true, the smoke and dust tracers for RRFS-SD are added.

4). The dry deposition velocity is an input variable to MYNN PBL.

5). The smoke and dust codes are significantly cleaned and improved.

### CODE REVIEW

File: .gitmodules
Comment (hu5970) [inline .gitmodules]: This should not point to your own branch and repository.

Comment (haiqinli) [reply .gitmodules]: It is updated to point to the NOAA-GSL repository and RRFS_dev branch.

**Comment by SamuelTrahanNOAA** [comment]
The branch fails to compile. I've encountered two errors so far.

1. A suite definition file is missing:

```
CRITICAL: Suite definition file suites/suite_FV3_HRRR_smoke.xml not found.
```

2. The HYCOM fails to compile, so you probably have out-of-date submodules at the ufs-weather-model level:

```
/scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/HYCOM-interface/HYCOM/mod_archiv.F90:1194:16:

 1194 |           'ssh', &
      |                1
Error: Different CHARACTER lengths (3/4) in array constructor at (1)
/scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/HYCOM-interface/HYCOM/mod_archiv.F90:1250:43:

 1250 |         call export_from_hycom_tiled(util2,cname_exp(k))  !can't use util1
      |                                           1
Error: Function 'cname_exp' at (1) has no IMPLICIT type
```

**Comment by SamuelTrahanNOAA** [comment]
Some baseline generation jobs are crashing:

1. control_csawmgt_debug and control_csawmg_debug have many lines like these:

```
77:
77: Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
77:
77: Backtrace for this error:
77: #0  0x2aceced273ff in ???
77: #1  0x27dfe3a in __tp_core_mod_MOD_pert_ppm
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/tp_core.F90:1201
77: #2  0x27f6206 in yppm
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/tp_core.F90:1043
77: #3  0x2812fae in __tp_core_mod_MOD_fv_tp_2d
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/tp_core.F90:202
77: #4  0x27029b5 in __fv_tracer2d_mod_MOD_tracer_2d_1l._omp_fn.4
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/fv_tracer2d.F90:270
77: #5  0x2acecdc96211 in GOMP_parallel
77:     at /tmp/Role.Apps/spack-stage/spack-stage-gcc-9.2.0-wqdecm4rkyyhejagxwmnabt6lscgm45d/spack-src/libgomp/parallel.c:171
77: #6  0x26e6b44 in __fv_tracer2d_mod_MOD_tracer_2d_1l
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/fv_tracer2d.F90:270
77: #7  0x2456c7d in __fv_dynamics_mod_MOD_fv_dynamics
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/model/fv_dynamics.F90:714
77: #8  0x22bbc40 in adiabatic_init
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90:1931
77: #9  0x231fb53 in __atmosphere_mod_MOD_atmosphere_init
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90:566
77: #10  0x20bc038 in __atmos_model_mod_MOD_atmos_model_init
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/atmos_model.F90:551
77: #11  0x1eabc09 in fcst_initialize
77:     at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/module_fcst_grid_comp.F90:756
```

2. cpld_debug_p8 has many errors like these:
```
121: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002b8be25a3818 ***
 88: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002b78c88ee7b8 ***
 44: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002b9b4bb557f8 ***
 90: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002b88f11ed7b8 ***
125: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002abcc8c2d818 ***
 48: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002af2e8fd37f8 ***
127: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002acc1fce8818 ***
 94: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002af0424927b8 ***
117: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002b56d21eb818 ***
 55: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002afb1557d7f8 ***
100: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002ae4a96527b8 ***
118: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/cpld_debug_p8/./fv3.exe': free(): invalid pointer: 0x00002ab9b79ef818 ***
```

3. The cpld_control_p8, cpld_control_c192_p8, cpld_bmark_p8, and cpld_control_c96_p8 are failing with messages like this, likely indicating that either the parm or input data are out-of-date:

```
122: At line 4376 of file /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/FV3/ccpp/physics/physics/module_mp_thompson.F90 (unit = 63, file = 'qr_acr_qgV2.dat')
122: Fortran runtime error: End of file
122:
122: Error termination. Backtrace:
```

**Comment by SamuelTrahanNOAA** [comment]
The errors so far make me suspect the fv3atm and ufs-weather-model are out-of-date.

**Comment by SamuelTrahanNOAA** [comment]
I'll hold off on running the Intel tests, which are numerous and expensive, until this is fixed. There's no point in using more cpu resources to test something we know isn't working.

**Comment by haiqinli** [comment]
The suite of FV3_HRRR_smoke is no longer existed, and it has been merged with FV3_HRRR. It is controlled by the namelist option of rrfs_sd to run as regular RRFS or RRFS with smoke and dust.

**Comment by SamuelTrahanNOAA** [comment]
> The suite of FV3_HRRR_smoke is no longer existed, and it has been merged with FV3_HRRR. It is controlled by the namelist option of rrfs_sd to run as regular RRFS or RRFS with smoke and dust.

You need to update the rt.conf and rt_gnu.conf.

**Comment by SamuelTrahanNOAA** [comment]
More tests have failed.

1. datm_cdeps_stochy_gefs

``
12: FATAL from PE     0: MPP_OPEN: error in OPEN for input.nml.
``

2. control_atm_aerosols fails with many lines like this:

```
 39: 
 39: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
 39: 
 39: #0  0x2b3d7f6343ff in ???
 39: #1  0x12f6475 in modeldatainitialize
 39:    at /scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/gnu-RRFS_dev_sd/GOCART/ESMF/UFS/Aerosol_Cap.F90:321
 ... many more lines of stack trace ...
```

3. control_c384gdas_wav deadlocks almost immediately, with this at the end of their PET files:

```
20220906 194516.308 INFO             PET111 ReadAttributes WAV_attributes:: end:
20220906 194516.308 INFO             PET111 ReadAttributes WAV_modelio:: start:
20220906 194516.308 INFO             PET111 ReadAttributes WAV_modelio:: end:
```

4. control_csawmgt and control_csawmg fail with heap corruption, spitting out errors like this:

```
141: *** Error in `/scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe': free(): invalid next size (fast): 0x0000000022da2ae0 ***
141: ======= Backtrace: =========
141: /lib64/libc.so.6(+0x81329)[0x2aed50625329]
141: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe[0x25baa18]
141: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe[0x25c3c9f]
141: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe[0x23feb4b]
141: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe[0x23878cb]
141: /scratch1/NCEPDEV/stmp2/Samuel.Trahan/FV3_RT/rt_74669/control_csawmgt/./fv3.exe[0x23856cb]
141: /apps/gnu/gcc-9.2.0/lib64/libgomp.so.1(GOMP_parallel+0x42)[0x2aed4f549212]
```

5. The regional_3km deadlocks at the end of the simulation. The last message to stderr is "post_fname=NATLEV.GrbF00"
6. The control_c192 and control_c384 behave similarly to regional_3km but their last print is a later forecast hour of the post.
7. The cpld_control_c384_p8 fails in the same way as the cpld_control_c192_p8

File: .gitmodules
Comment (hu5970) [reply .gitmodules]: This is a mistake. It should not point to NOAA-GSL until the PR is merged into NOAA-GSL. Sorry for confusion.

**Comment by hu5970** [comment]
@SamuelTrahanNOAA Those regression tests are for NOAA-GSL/RRFS_dev branch or they are for NOAA-GSL/develop branch?

**Comment by SamuelTrahanNOAA** [comment]
> @SamuelTrahanNOAA Those regression tests are for NOAA-GSL/RRFS_dev branch or they are for NOAA-GSL/develop branch?

I think they were merged incorrectly, so they're a mix of the old RRFS_dev ones (which should work) and develop ones (which may not work).

**Comment by hu5970** [comment]
@SamuelTrahanNOAA Haiqing's work is to add the smoke/dust on top of the RRFS_dev. If you are not turning on smoke/dust, it will repeat the current RRFS_dev. I made a mistake to ask Haiqing to change the submodule fv3lam to point to GSL version. It should point to haiqing's fv3lam until the fv3lam merged.

File: .gitmodules
Comment (hu5970) [reply .gitmodules]: @haiqinli Canyou switch this back? thanks.

**Comment by SamuelTrahanNOAA** [comment]
That would explain many of the failures, but not all of them. The data atmosphere test failed, which does not use fv3atm, so some of the ufs-weather-model submodules or regression tests are incorrect. The HYCOM, WW3, and MOM6 are either failing to build or run, which points to the same issues (submodules & regression tests).

Perhaps you shouldn't use his ufs-weather-model either?

File: .gitmodules
Comment (SamuelTrahanNOAA) [reply .gitmodules]: @hu5970 It would be better if you made your own PRs at the ufs-weather-model and fv3atm level, pointing to @haiqinli's  underlying ccpp-physics.

**Comment by SamuelTrahanNOAA** [comment]
~~The problem is not with this branch; the RRFS_dev is already failing the regression tests. I think there are a few files missing or incorrect in the tests/parm directory. I'm fixing this now, and I'll let you know the results. Once Hera is back, I can tell for certain since I'll have the directory where I ran the final regression tests for the past few RRFS_dev PRs.~~

EDIT: The RRFS_dev passes regression tests; the problem is with this branch.

**Comment by SamuelTrahanNOAA** [comment]
After my updates to this branch's regression test system, all tests match previous baselines except these:

- rrfs_smoke_conus13km_hrrr_warm (expected to change)
- rap_rrtmgp

Are the rap_rrtmgp results supposed to change?

**Review by SamuelTrahanNOAA — CHANGES_REQUESTED** [review]
1. rap_rrtmgp test fails
2. FV3/ccpp/physics hash is out of date with ccpp-physics PR

File: .gitmodules
Comment (SamuelTrahanNOAA) [reply .gitmodules]: There is nothing wrong with pointing to your own repository. Many people do that, and it is corrected during the commit process.

What's most important is that the hashes be up-to-date, which they aren't right now. The FV3/ccpp/physics differs from the top of the ccpp-physics PR

**Comment by SamuelTrahanNOAA** [comment]
Curiously, the rap_rrtmgp_debug succeeds, but the rap_rrtmgp fails. That means the unoptimized case matches its baseline, but the optimized one has changed results.

HERA: /scratch2/BMC/wrfruc/Samuel.Trahan/for-haiqin/failing-rap-rrtmgp

**Comment by SamuelTrahanNOAA** [comment]
This branch crashes in debug mode and crashes during restart. I have a fix for the former problem, and I'm working on the later. My fixes will go to the authoritative repository first, and then I'll help @haiqinli back-port it to this older version of the code.

**Comment by SamuelTrahanNOAA** [comment]
@haiqinli Please update your branches; I've merged the boundary and lsm ruc bugfix PR to RRFS_dev.

**Comment by haiqinli** [comment]
@SamuelTrahanNOAA Thanks. I will update my branches to include your changes.

**Review by SamuelTrahanNOAA — APPROVED** [review]

**Pull request merged**
